File: src/main/java/ting/leetcode/java/linkedlist/AddTwoNumbers.java
package ting.leetcode.java.linkedlist;

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode pointer = dummy;
        int overflow = 0;

        while (l1 != null || l2 != null) {
            int sum = 0;
            if (l1 != null && l2 != null) {
                sum = l1.val + l2.val + overflow;
            } else if (l2 == null) {
                sum = l1.val + overflow;
            } else {
                sum = l2.val + overflow;
            }

            overflow = sum / 10;
            ListNode newNode = new ListNode(sum % 10);
            pointer.next = newNode;
            pointer = newNode;
            if (l1 != null) {
                l1 = l1.next;
            }
            if (l2 != null) {
                l2 = l2.next;
            }
        }

        if (overflow != 0) {
            pointer.next = new ListNode(overflow);
        }

        return dummy.next;
    }
}

File: kotlin-solution/src/leetcode/linkedlist/AddTwoNumbers.kt
package leetcode.linkedlist

class AddTwoNumbers {
    fun addTwoNumbers(l1: leetcode.linkedlist.ListNode?, l2: leetcode.linkedlist.ListNode?): leetcode.linkedlist.ListNode? {
        val customHead = leetcode.linkedlist.ListNode(0)
        var currentNode: leetcode.linkedlist.ListNode? = customHead
        var firstPointer = l1
        var secondPointer = l2
        while (firstPointer != null || secondPointer != null) {
            val l1nodeValue = firstPointer?.`val` ?: 0
            val l2nodeValue = secondPointer?.`val` ?: 0
            currentNode?.next = leetcode.linkedlist.ListNode(l1nodeValue + l2nodeValue).also {
                firstPointer = firstPointer?.next
                secondPointer = secondPointer?.next

                if (it.`val` >= 10) {
                    when(firstPointer) {
                        null -> {
                            it.next = leetcode.linkedlist.ListNode(1)
                            firstPointer = it.next
                        }
                        else -> firstPointer?.`val` = firstPointer?.`val`!! + 1
                    }
                    it.`val` -= 10
                }
                currentNode = it
            }
        }
        return customHead.next
    }
}
